spanner/row.go

```diff
@@ -378,3 +386,133 @@ func (r *Row) ToStructLenient(p interface{}) error {
 		true,
 	)
 }
+
+// SelectAll scans rows into a slice (v)
+func SelectAll(rows Iterator, v interface{}, options ...DecodeOptions) error {
+	if rows == nil {
+		return fmt.Errorf("rows is nil")
+	}
+	if v == nil {
+		return fmt.Errorf("p is nil")
+	}
+	vType := reflect.TypeOf(v)
+	if k := vType.Kind(); k != reflect.Ptr {
+		return errToStructArgType(v)
+	}
+	sliceType := vType.Elem()
+	if reflect.Slice != sliceType.Kind() {
+		return errNotASlicePointer()
+	}
+	sliceVal := reflect.Indirect(reflect.ValueOf(v))
+	itemType := sliceType.Elem()
+	s := &decodeSetting{}
+	for _, opt := range options {
+		opt.Apply(s)
+	}
+
+	isPrimitive := itemType.Kind() != reflect.Struct
+	var pointers []interface{}
+	isFistRow := true
+	rowIndex := -1
+	return rows.Do(func(row *Row) error {
+		sliceItem := reflect.New(itemType).Elem()
+		if isFistRow {
+			nRows := rows.RowsReturned()
+			if nRows != -1 {
+				sliceVal = reflect.MakeSlice(sliceType, int(nRows), int(nRows))
+				reflect.ValueOf(v).Elem().Set(sliceVal)
+				rowIndex++
+			}
+			if isPrimitive {
+				if len(row.fields) > 1 {
+					return errTooManyColumns()
+				}
+				pointers = []interface{}{sliceItem.Addr().Interface()}
+			} else {
+				var err error
+				if pointers, err = structPointers(sliceItem, row.fields, s.Lenient); err != nil {
+					return err
+				}
+			}
+			isFistRow = false
+		}
+		if len(pointers) == 0 {
+			return nil
+		}
+		err := row.Columns(pointers...)
+		if err != nil {
+			return err
+		}
+		for i, p := range pointers {
+			if p == nil {
+				continue
+			}
+			sliceItem.Field(i).Set(reflect.ValueOf(p).Elem())
+		}
+		if rowIndex >= 0 {
+			sliceVal.Index(rowIndex).Set(sliceItem)
```

Comment:
Yes, using `reflect.MakeSlice(sliceType, 0, int(nRows))` was 25% slower compared to preallocated and setting the value via index, I have updated the code to not panic if `RowsReturned ` is lower bound, 
btw it will be lower bound only for DML and PDML so normal Select queries should work needing append.